runtime/onert/core/src/odc/CodegenLoader.h

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (c) 2024 Samsung Electronics Co., Ltd. All Rights Reserved
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef __ONERT_ODC_CODEGEN_LOADER_H__
+#define __ONERT_ODC_CODEGEN_LOADER_H__
+
+#include "odc/ICodegen.h"
+
+#include <functional>
+#include <memory>
+
+namespace onert
+{
+namespace odc
+{
+
+/**
+ * @brief Class to manage loading and unloading of dynamic library containing
+ *        implementation of ICodegen interface.
+ */
+class CodegenLoader
+{
+public:
+  /**
+   * @brief Typedef for function pointer to destroy loaded library handle
+   */
+  using dlhandle_destroy_t = std::function<void(void *)>;
+  /**
+   * @brief Typedef for function pointer to create instance of ICodegen
+   */
+  using factory_t = ICodegen *(*)();
+  /**
+   * @brief Typedef for function pointer to destroy instance of ICodegen
+   */
+  using codegen_destory_t = void (*)(ICodegen *);
+
+  /**
+   * @brief Get singleton instance of CodegenLoader
+   * @return Reference to singleton instance of CodegenLoader
+   */
+  static CodegenLoader &instance();
+
+private:
+  // cannot create instance of CodegenLoader outside of this class
+  CodegenLoader() = default;
+  CodegenLoader(CodegenLoader const &) = delete;
+  CodegenLoader &operator=(CodegenLoader const &) = delete;
```

Comment:
Why don't you move these copy operations in `public`?

Reply:
This class is a singleton object. So I disabled the copy operators. This ensures one cannot create a copy of ICodegen after it is returned by get() function of ICodegen.

And even if it is deleted in the declaration, because the copy operators are declared, move operators are not implicitly declared as you said in todays's seminar.😄

Reply:
> This class is a singleton object. So I disabled the copy operators. This ensures one cannot create a copy of ICodegen after it is returned by get() function of ICodegen.

You already declared the functions as _Deleted functions_. If so, is there any reason to declare them in `private`? I don't know what happen when _Deleted functions_ are declared in `private`.

> And even if it is deleted in the declaration, because the copy operators are declared, move operators are not implicitly declared as you said in todays's seminar.😄

I wasn't worried about that .😄

Reply:
As we saw offline, Effective Modern C++ recommends that deleted functions are declared public, not private for better error messages.

--- 
The mention In the book,

By convention, deleted functions are declared public, not private. There’s a reason for that. When client code tries to use a member function, C++ checks accessibility before deleted status. When client code tries to use a deleted private function, some compilers complain only about the function being private, even though the function’s accessibility doesn’t really affect whether it can be used. It’s worth bearing this in mind when revising legacy code to replace private-and-not-defined member functions with deleted ones, because making the new functions public will generally
result in better error messages.

Reply:
@ragmani 
I saw the same phrases in the book. It would be better to declare delete function in public. So I updated this PR.